hudi-common/src/main/java/org/apache/hudi/metadata/HoodieTableMetadataUtil.java

```diff
@@ -2550,9 +2578,20 @@ public static HoodieData<HoodieRecord> convertMetadataToPartitionStatsRecords(Ho
       return engineContext.parallelize(partitionedWriteStats, parallelism).flatMap(partitionedWriteStat -> {
         final String partitionName = partitionedWriteStat.get(0).getPartitionPath();
         // Step 1: Collect Column Metadata for Each File part of current commit metadata
-        List<List<HoodieColumnRangeMetadata<Comparable>>> fileColumnMetadata = partitionedWriteStat.stream()
+        List<List<HoodieColumnRangeMetadata<Comparable>>> rawFileColumnMetadata = partitionedWriteStat.stream()
             .map(writeStat -> translateWriteStatToFileStats(writeStat, dataMetaClient, validColumnsToIndex, tableSchema))
             .collect(Collectors.toList());
+        // convert to payload and reconstruct the stats to maintain parity for certain data types where avro wrapping and unwrapping could change the types.
+        List<List<HoodieColumnRangeMetadata<Comparable>>> fileColumnMetadata = rawFileColumnMetadata.stream()
+            .map(new Function<List<HoodieColumnRangeMetadata<Comparable>>, List<HoodieColumnRangeMetadata<Comparable>>>() {
+              @Override
+              public List<HoodieColumnRangeMetadata<Comparable>> apply(List<HoodieColumnRangeMetadata<Comparable>> hoodieColumnRangeMetadata) {
+                return HoodieMetadataPayload.createColumnStatsRecords(partitionName, hoodieColumnRangeMetadata, false).map(record -> {
+                  return HoodieColumnRangeMetadata.fromColumnStats((((HoodieMetadataPayload) record.getData()).getColumnStatMetadata().get()));
+                }).collect(toList());
+              }
+            }).collect(toList());
```

Comment:
so we are converting from a list of HoodieColumnRangeMetadata to HoodieMetadataPayload and then back to HoodieColumnRangeMetadata. Sounds like a lot of work. Can we avoid repeated transformations? My suggestions is to handle parity in `HoodieTableMetadataUtil.collectColumnRangeMetadata`, which is used even in the append handle.

Reply:
As I was telling you, I will try to avoid full payload serde.